% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/mixfine.R
\name{mixfine}
\alias{mixfine}
\title{Combining total and allele-specific reads for fine-mapping}
\usage{
mixfine(
  geno1,
  geno2,
  y1,
  y2,
  ytotal,
  lib_size,
  cov_offset = NULL,
  trc_cutoff = 20,
  asc_cutoff = 5,
  weight_cap = 100,
  asc_cap = 5000
)
}
\arguments{
\item{geno1}{genotype of haplotype 1 (dimension = N x P)}

\item{geno2}{genotype of haplotype 2 (dimension = N x P)}

\item{y1}{allele-specific read count for haplotype 1 (dimension = N x 1)}

\item{y2}{allele-specific read count for haplotype 2 (dimension = N x 1)}

\item{ytotal}{total read count (dimension = N x 1)}

\item{lib_size}{library size (dimension = N x 1)}

\item{cov_offset}{predicted effect of covariates on total read count response term,
log(ytotal / lib_size / 2) (dimension = N x 1)}

\item{trc_cutoff}{total read count cutoff to exclude observations with ytotal lower than the cutoff}

\item{asc_cutoff}{allele-specific read count cutoff to exclude observations with y1 or y2 lower than asc_cutoff}

\item{weight_cap}{the maximum weight difference (in fold) is min(weight_cap, floor(sample_size / 10)). The ones exceeding the cutoff is capped. Set to NULL then no weight_cap applied.}

\item{asc_cap}{exclude observations with y1 or y2 higher than asc_cap}
}
\value{
fine-mapping results (95% credible set and PIP)
}
\description{
Given total and allele-specific read counts along with library size,
the two-step inference procedure is implemented to perform fine-mapping.
Step 1: estimate variances of total and allele-specific count response respectively
along with intercept for total count response
Step 2: transform observation according to step 1 and perform fine-mapping
using susieR::susie
}
\examples{
mixfine(
  geno1 = matrix(sample(c(0, 0.5, 1), 200, replace = TRUE), ncol = 2),
  geno2 = matrix(sample(c(0, 0.5, 1), 200, replace = TRUE), ncol = 2),
  y1 = rpois(100, 100),
  y2 = rpois(100, 100),
  ytotal = rpois(100, 1000),
  lib_size = rpois(100, 10000),
  cov_offset = runif(100),
  trc_cutoff = 100,
  asc_cutoff = 50,
  weight_cap = 100,
  asc_cap = 1000
)

}
